src/java.base/share/classes/jdk/internal/event/ThrowableTracer.java

```diff
@@ -22,47 +22,50 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
+package jdk.internal.event;
 
-package jdk.jfr.internal.instrument;
-
+import java.lang.reflect.Field;
 import java.util.concurrent.atomic.AtomicLong;
 
-import jdk.jfr.events.EventConfigurations;
-import jdk.jfr.events.ErrorThrownEvent;
-import jdk.jfr.events.ExceptionThrownEvent;
-import jdk.jfr.internal.event.EventConfiguration;
-
+/**
+ * Helper class for exception events.
+ */
 public final class ThrowableTracer {
 
     private static final AtomicLong numThrowables = new AtomicLong();
 
-    public static void traceError(Error e, String message) {
-        if (e instanceof OutOfMemoryError) {
+    public static void enable() throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+        Field field = Throwable.class.getDeclaredField("tracing");
+        field.setAccessible(true);
+        field.setBoolean(null, true);
+    }
+
+    public static void traceError(Class<?> clazz, String message) {
+        if (OutOfMemoryError.class.isAssignableFrom(clazz)) {
```

Comment:
StackOverflowError is likely problematic too, maybe it should be VirtualMachineError.

Reply:
I remember asking the same question ten years ago, when Nils did the original implementation, but I think it was only needed for OOM, because it creates an infinite loop when the event object was allocated, which resulted in a StackOverflowError instead OOM.

Some OOM tests failed with JFR enabled.

The event object allocation has been removed, but I think we can run into the allocation recursion by other means. I looked into it a few years ago, but I don't remember exactly why it failed.

If a SOE happens, I think we are fine. There is something that prevents infinite recursion when the SOE object is created. Perhaps it is preallocated? I prefer to not change the behavior, at least not in this PR.

Reply:
I filed an issue to investigate if there is problem with SOE, or if the OOM check is really needed now.
https://bugs.openjdk.org/browse/JDK-8319579

Reply:
> I filed an issue to investigate if there is a problem with SOE, or if the OOM check is really needed now. https://bugs.openjdk.org/browse/JDK-8319579
> 
> Regardless of outcome, It would be good to document the results of the investigation in the code.

Okay, the scenario I was wondering about is where traceError with SOE and whether that could trigger a loop too.

Reply:
I understand.